lantiq: kernel: xway-nand: Move ECC engine setting to new DT binding

kestrel1974 opened the pull request:
Hi,

this is prework to adding support for another device. I am working with the kernel maintainers, but its hard to come up with something without owning another lantiq device with NAND, e.g. the BT Homehub 5a.
I would be great, if someone owning e.g. a BT Homehub 5a could test if the device tree changes and removing the hardcoded setting in the linux kernel work for them. I will then send the kernel patches upstream.
I have searched the kernel sources, but I am not sure if the DT setting  nand-ecc-mode = "soft" is overwritten by some chip driver and like to make sure that this change does not break something.

Thanks.

When trying to add support for another device with Micron NAND chips,
it was discovered that the default setting in the kernel source does
not work with Micron Chips, since the device trees setting is
overwritten and hard coded by the kernel xway_nand driver.
This patch sets the ECC mode to soft in the device tree for the
devices that currently work fine with the hard coded setting and the
hardcoded setting is removed in the kernel driver.

Signed-off-by: Daniel Kestrel <kestrel1974@t-online.de>

kestrel1974 committed:
lantiq: kernel: xway-nand: Move ECC engine setting to new DT binding

When trying to add support for another device with Micron NAND chips,
it was discovered that the default setting in the kernel source does
not work with Micron Chips, since the device trees setting is
overwritten and hard coded by the kernel xway_nand driver. This was
the original reason for this PR.
A kernel patch sets the default ECC mode to soft without overwriting
the device tree settings and the device tree for devices using it
are updated with new parameters because the old ones are deprecated
by torvalds/linux@533af69.
A patch for kernel 5.4 is provided to support the new settings
because kernel 5.4 does not support it.

Signed-off-by: Daniel Kestrel <kestrel1974@t-online.de>

abajk commented:
Hi,
Everything seems to be working. Tested on 5.4[1] and 5.10[2]. 

Tested-by: Aleksander Jan Bajkowski <olek2@wp.pl> # tested on BT Home Hub 5A

1. https://pastebin.com/7cFUvpDM
2. https://pastebin.com/SpkGtkSK

kestrel1974 commented:
@abajk Thanks for testing. There are a couple of currently supported devices which have nand-ecc-mode set to "on-die" but due to the hard coded settings have effectively used nand-ecc-mode "soft". Do you suggest to change those devices properties from "on-die" to "soft" too?

abajk commented:
> 
> 
> @abajk Thanks for testing. There are a couple of currently supported devices which have nand-ecc-mode set to "on-die" but due to the hard coded settings have effectively used nand-ecc-mode "soft". Do you suggest to change those devices properties from "on-die" to "soft" too?

The Fritzbox 3370 and 7362sl use the Micron MT29F1G08ABADAWP memory chips. They also have an ecc block built in [1]. I wouldn't change the "on-die" property.

1. https://eu.mouser.com/datasheet/2/671/micron_technology_micts06228-1-1759217.pdf

abajk commented:
nand-ecc-mode is deprecated since 5.11[1].

1. https://github.com/torvalds/linux/commit/533af69cf1a2a6bff211d2abe44044980cc23602

kestrel1974 commented:
@abajk Should I add both to the device tree and remove nand-ecc-mode when openwrt moves away from kernel 5.10 ???

hauke commented:
It looks like nand-ecc-engine is already available in kernel 5.10, see here: https://github.com/torvalds/linux/commit/a8c7ffdb5fdde3a57c0b654f66f4d81325abe69f

I would prefer if you use this for kernel 5.10, then we do not have to change this again when we upgrade the kernel.

kestrel1974 commented:
@hauke Ok, but when the kernel patch is accepted for 5.10 and 5.4, then trunk will be broken as long as the default kernel is 5.4?! Or should I create an openwrt patch for 5.4 that adds nand-ecc-engine?

adschm commented on target/linux/lantiq/patches-5.10/0019-MTD-nand-remove-hardcoded-soft-ecc.patch:
```diff
@@ -0,0 +1,11 @@
+--- a/drivers/mtd/nand/raw/xway_nand.c
++++ b/drivers/mtd/nand/raw/xway_nand.c
```
I think these patches should have a header to explain what they do.

hauke replied:
I agree with @adschm  it would be nice to have a title why this patch is needed.

hauke commented:
Sorry I did thought about that we share the dts files between kernel 5.4 and 5.10, but your solution looks good. I agree with @adschm  please add a description to both patches.

hvegh reviewed (approved):
Tested using this branch:

https://github.com/hvegh/openwrt/commit/de539710bd4df6af811acb1f23fcdc20a7d9b41a

janh commented:
I recently opened a bug report about this, before I found this pull request: https://bugs.openwrt.org/index.php?do=details&task_id=4035

I can confirm that this fixes the NAND ECC issues with kernel 5.10 on a Fritzbox 7372 SL. It also works with kernel 5.4 and on a Fritzbox 7412.

However, I am unsure if this is the best fix for the upstream kernel. I wrote a more detailed message about this as a reply to your upstream patch.

hvegh commented:
> I recently opened a bug report about this, before I found this pull request: https://bugs.openwrt.org/index.php?do=details&task_id=4035
> 
> I can confirm that this fixes the NAND ECC issues with kernel 5.10 on a Fritzbox 7372 SL. It also works with kernel 5.4 and on a Fritzbox 7412.
> 
> However, I am unsure if this is the best fix for the upstream kernel. I wrote a more detailed message about this as a reply to your upstream patch.

My best guess is that if you write the kernel partition with software ECC (for example after booting ramfs image)  and the bootloader uses hw ECC, and both differ, you get problems...

Does the DTS file correctly specify  hw ECC mode for this device?

janh commented:
> My best guess is that if you write the kernel partition with software ECC (for example after booting ramfs image) and the bootloader uses hw ECC, and both differ, you get problems...

Yes, this is excactly what happens. I did't know the exact cause when I wrote the bug report, but after finding this pull request and looking at the relevant kernel commits, it is clear why this is happening.

> Does the DTS file correctly specify hw ECC mode for this device?

The DTS file is correct. It worked with kernel 5.4, but is broken in 5.10, because the value from the device tree is now overwritten in `xway_attach_chip` (which is fixed by this change).

Here is a more detailed description of the kernel changes that resulted in this issue: https://lore.kernel.org/linux-mtd/51f2ebf4-6df1-eba5-99f1-1ec88e475d20@3e8.eu/

kestrel1974 commented:
@janh Do you have any suggestion? I thought if the [patch 2] is added for NAND, I could add two dtbs (similar to 3370) one for Micron and one for non Micron NAND. The procedure to figure out which NAND your Fritzbox has would be to generate Support Data and then look in the Text file for the dmesg output and then use the appropriate initramfs (either the Micron or non-Micron) image to do it. Also [see the pull request](https://github.com/openwrt/openwrt/pull/4426), you might want to comment there.
At least for all the NAND devices, we could fix the DTBs as in the pull request and nothing will break?!

hvegh commented:
> > My best guess is that if you write the kernel partition with software ECC (for example after booting ramfs image) and the bootloader uses hw ECC, and both differ, you get problems...
> 
> Yes, this is excactly what happens. I did't know the exact cause when I wrote the bug report, but after finding this pull request and looking at the relevant kernel commits, it is clear why this is happening.
> 
> > Does the DTS file correctly specify hw ECC mode for this device?
> 
> The DTS file is correct. It worked with kernel 5.4, but is broken in 5.10, because the value from the device tree is now overwritten in `xway_attach_chip` (which is fixed by this change, although it removes the default value entirely and thus requires DTS changes for all devices that relied on that default).
> 
> Here is a more detailed description of the kernel changes that resulted in this issue: https://lore.kernel.org/linux-mtd/51f2ebf4-6df1-eba5-99f1-1ec88e475d20@3e8.eu/

> This is fixed in d525914b5bd8 ("mtd: rawnand: xway: Move the ECC
> initialization to ->attach_chip()"). However, this makes it impossible
> to configure the ECC engine type in the device tree, as it is now
> overwritten by the default value in xway_attach_chip.

I have not yet had the time to look into the whole flow of events involved in mounting a flash device. Ideally I would imagine that the low level NAND drivers would select sensible defaults based on geometry and capabilities of the device, unless overruled by device tree configuration.

Maybe a clear improvement would be that the kernel actually tells us what is being used for ECC.... 
```
[    0.681086] nand: device found, Manufacturer ID: 0x2c, Chip ID: 0xdc
[    0.686032] nand: Micron MT29F4G08ABADAWP
[    0.690164] nand: 512 MiB, SLC, erase size: 128 KiB, page size: 2048, OOB size: 64
```

janh commented:
@kestrel1974: I also don't know of a good way for supporting different NAND chips with a single device tree. So, adding two distinct devices (similar to Fritzbox 3370) seems to be the most realistic option.

As for your patch, the currently hard-coded ECC engine type definitely needs fixing. And from my testing with Fritzbox 7412 and 7362 SL your pull request seems to work for both devices with both kernel versions. I just don't think this is the best solution for upstreaming.

> Also [see the pull request](https://github.com/openwrt/openwrt/pull/4426), you might want to comment there.

That links to this pull request. Did you mean to link somewhere else?

kestrel1974 commented:
@janh sorry for the wrong link, I just thought its another comment on the fritzbox 7490 work.

hauke commented:
Please add an empty new line after the git commit title.

janh commented:
This issue is in the process of being fixed upstream: https://lore.kernel.org/linux-mtd/20210927183150.4be87140@xps13/
The upstream change is going to fix configuration of on-die ECC in the device tree, while keeping software ECC as default (i.e. no DTS changes are necessary on kernel 5.10).

Thus it probably doesn't make sense to apply this pull request as-is. Although the part about switching from `nand-ecc-mode` to `nand-ecc-engine` is still going to be necessary to support newer kernels in the future.

janh commented:
The proposed upstream fix is now on the linux-mtd mailing list:
https://lore.kernel.org/linux-mtd/20210928222258.199726-11-miquel.raynal@bootlin.com/T/

kestrel1974 commented:
@janh Thanks for looking into that. The patch for 5.10 was never intended to be merged, just something like a test for a possible patch. So can I remove the patch for 5.10 from this PR?
2nd question, but the micron and non-micron DTB (as for fritzbox 3370) would still be required?
Do the models that you mentioned/tested also have manufacture lots with micron and non-micron nand or do all fritzboxes need two DTBs and two device entries?
In the wiki for 3370 there is a process described from the AVM stock firmware to generate a text document that contains the complete dmesg of the stock firmware, which includes the type of NAND in the device, maybe that needs to be added to all models.

hauke commented:
It probably takes ~3 weeks till this patch shows up in kernel 5.10.X as a stable update, we should cherry pick it. Do you want to update this pull request with the change or create a separate one?

kestrel1974 commented:
@hauke I have removed the 5.10 patch, changed the 5.4. patch to have the same changes as the 5.10 patch, added a short description for the 5.4 patch added a line break in the commit message.

janh commented:
> I have removed the 5.10 patch, changed the 5.4. patch to have the same changes as the 5.10 patch,

The issue of the hardcoded ECC engine is a regression in kernel 5.10, so for 5.4, no change to xway_nand is needed. (You can also see this from the commit message of the patch on the linux-mtd mailing list. The "Fixes" tag specifies that it is a fix for an issue in that commit. Thus, the fix only applies to kernel versions that actually contain that commit.)

Also, explicitly configuring software ECC in all the DTS files shouldn't be necessary, as that is already the default (it was required with your originally proposed fix).

kestrel1974 commented:
@janh I have removed the part of the 5.4 patch that changes the xway nand driver, but have left the part that allows to use the new device tree API.
Defaults can change and so I think its better to leave the values telling the code what is wanted and not rely on defaults.

hauke commented:
Thank you for the patch, I applied it to master and also applied the upstream fix in 6bb51c9d1cd5eec132e6017563fb2c35d026e7cf
